### ISSUE TITLE
Cannot complete terraform init for 3.x "hashicorp/aws" provider

### ISSUE THREAD

**Issue by minjas** [hashicorp/terraform#30605]
### Terraform Version

```
v1.1.2
```

### Terraform Configuration Files

```terraform
terraform {
  required_version = ">= 0.15"
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "~> 3.0"
    }
  }
}
```

### Expected Behavior

It should complete terraform init or terraform init -upgrade command successfully 

### Actual Behavior

_**│ Error: Failed to query available provider packages
│
│ Could not retrieve the list of available versions for provider hashicorp/aws: no available releases match the given
│ constraints >= 2.0.0, >= 2.48.0, >= 2.53.0, >= 3.0.0, ~> 3.0, >= 3.62.0, >= 3.63.0, >= 3.67.0, >= 4.0.0**_

This is existing project, worked one month ago with same code base. It suddenly stopped working on **terraform init** showing message above.

I tried to delete .terraform folder and .lock files and run terraform init --upgrade but got same message above.

I used same **version = "~> 3.0"** in every config file.

It starts happening also on other PCs for other team members.

### Steps to Reproduce

terraform init
or
terraform init -upgrade

on 3.x provider versions in project.

### Additional Context

It is atypical it stopped working and nothing was changed in repo. Same configuration worked fine one month ago. It only works if I use version = "~> 4.0", but then terraform plan fails because of braking changes n 4.x version. Want to avoid fixing it at the moment.

**Comment by crw** [comment]
Hi @minjas, this answer comes via my colleague @justinretzolk. 

The incompatibility is between the ` ~> 3.0` and ` >= 4.0.0`. 
` ~> 3.0` restricts to a maximum version that starts with `3`
` >= 4.0.0` restricts to a minimum version that starts with `4`
This is thus impossible to resolve. For more info on interpreting version constraints, see https://www.terraform.io/language/expressions/version-constraints

It seems likely that you are referencing a module that has the `>=4.0.0` constraint. Could you please check on that and let me know what you find? I'll keep this issue open for now just in case it does turn out to be a bug.

**Comment by minjas** [comment]
Hi @crw ,

I double checked my files, there is no 4.x module reference anywhere.

I'm pasting the whole command error output. 

**Initializing provider plugins...
- Finding hashicorp/template versions matching ">= 2.0.0"...
- Finding hashicorp/local versions matching ">= 1.3.0"...
- Finding hashicorp/aws versions matching ">= 2.0.0, >= 2.48.0, >= 2.53.0, >= 3.0.0, ~> 3.0, >= 3.62.0, >= 3.63.0, >= 3.67.0, >= 4.0.0"...
- Finding hashicorp/random versions matching ">= 3.1.0"...
- Installing hashicorp/template v2.2.0...
- Installed hashicorp/template v2.2.0 (signed by HashiCorp)
- Installing hashicorp/local v2.1.0...
- Installed hashicorp/local v2.1.0 (signed by HashiCorp)
- Installing hashicorp/random v3.1.0...
- Installed hashicorp/random v3.1.0 (signed by HashiCorp)
╷
│ Error: Failed to query available provider packages
│
│ Could not retrieve the list of available versions for provider hashicorp/aws: no available releases match the given
│ constraints >= 2.0.0, >= 2.48.0, >= 2.53.0, >= 3.0.0, ~> 3.0, >= 3.62.0, >= 3.63.0, >= 3.67.0, >= 4.0.0**


Basically I have it configured in workspaces\development\terraform.versions.tf file, do you have suggestion where else I should have it configured?

As I said before it is strange this starts happening also on other projects without changing anything related to versions. And other team members started having this after deleting .terraform project.

**Comment by minjas** [comment]
Just update, if I replace "~> 3.0" with "~> 4.0" it works. But, as I said, I have to update my files so terraform plan can work.

**Comment by minjas** [comment]
After upgrading to ~> 4.0 new .terraform folder is created and my .lock file shows this 

```
provider "registry.terraform.io/hashicorp/aws" {
  version     = "4.3.0"
  constraints = ">= 2.0.0, >= 2.48.0, >= 2.53.0, >= 3.0.0, >= 3.62.0, >= 3.63.0, >= 3.67.0, >= 4.0.0, ~> 4.0"
  hashes = [
    "h1:+z4+nEWPJ7NGu4Zo2U/3ghgN3+M3tckgUiTZI0c8k9I=",
    "zh:087c67e5429f343a164221c05a83f152322f411e7394f8a39ed81a75982af1f2",
    "zh:2e852a1b107e5324524874e1cd98bcf3a69284b4fe04750aa373054177c54214",
    "zh:4b9a54b5895f945827832e6ddd16ff107301fedf47acbd83d17d4e18bbf10bb1",
    "zh:64dfc02bc85f5df2f51ff942fc78d72fcd0db17b0f53e1fae380e58adbd239b3",
    "zh:766f9aef619cfd23e924aee523791acccd30b6d8f1cc0ed1a7b5c953bf8c5392",
    "zh:90048d87ff3071a4356cf91916b46a7ec69ba55bcba5765b598d3fe545d4c6ca",
    "zh:c51f5b238af37c63e9033a12fd7fedc87c03eb966f5f5c7786eb6246e8bf3071",
    "zh:d0df94d3112a25de609dfb55c5e3b0d119dea519a2bdd8099e64a8d63f22b683",
    "zh:de166ecfeed70f570cea72ec094f00c2f997496b3226fa08518e7cd4a73884e1",
    "zh:e31c31d00f42ea2dbaab1ad4c245da5cfff63e28399b5a5795b5e6a826c6c8af",
    "zh:f93725afd8410194ede51d83505327aa1ae6a9b4280cf31db649c62c7dc203ae",
  ]
}
```

when I run terraform plan after this, I get attribute errors inside .terraform folder. According to documentation this is old syntax for 4.x version and I'm confused why it is generated inside .terraform folder as it is during "hashicorp/aws" module upgrade to 4.x.

> │ Warning: Argument is deprecated
│
│   with module.ecs.aws_ecs_cluster.this,
│   on .terraform\modules\ecs\main.tf line 1, in resource "aws_ecs_cluster" "this":
│    1: resource "aws_ecs_cluster" "this" {
│
│ Use the aws_ecs_cluster_capacity_providers resource instead
│
│ (and one more similar warning elsewhere)
╵
╷
│ Error: Value for unconfigurable attribute
│
│   with module.terraform_state_backend.aws_s3_bucket.default,
│   on .terraform\modules\terraform_state_backend\main.tf line 105, in resource "aws_s3_bucket" "default":
│  105: resource "aws_s3_bucket" "default" {
│
│ Can't configure a value for "logging": its value will be decided automatically based on the result of applying this
│ configuration.
╵
╷
│ Error: Value for unconfigurable attribute
│
│   with module.terraform_state_backend.aws_s3_bucket.default,
│   on .terraform\modules\terraform_state_backend\main.tf line 105, in resource "aws_s3_bucket" "default":
│  105: resource "aws_s3_bucket" "default" {
│
│ Can't configure a value for "versioning": its value will be decided automatically based on the result of applying
│ this configuration.
╵
╷
│ Error: Value for unconfigurable attribute
│
│   with module.terraform_state_backend.aws_s3_bucket.default,
│   on .terraform\modules\terraform_state_backend\main.tf line 105, in resource "aws_s3_bucket" "default":
│  105: resource "aws_s3_bucket" "default" {
│
│ Can't configure a value for "replication_configuration": its value will be decided automatically based on the result
│ of applying this configuration.
╵
╷
│ Error: Value for unconfigurable attribute
│
│   with module.terraform_state_backend.aws_s3_bucket.default,
│   on .terraform\modules\terraform_state_backend\main.tf line 105, in resource "aws_s3_bucket" "default":
│  105: resource "aws_s3_bucket" "default" {
│
│ Can't configure a value for "server_side_encryption_configuration": its value will be decided automatically based on
│ the result of applying this configuration.
╵
╷
│ Error: Value for unconfigurable attribute
│
│   with module.terraform_state_backend.aws_s3_bucket.default,
│   on .terraform\modules\terraform_state_backend\main.tf line 109, in resource "aws_s3_bucket" "default":
│  109:   acl           = var.acl
│
│ Can't configure a value for "acl": its value will be decided automatically based on the result of applying this
│ configuration.
╵
╷
│ Error: Value for unconfigurable attribute
│
│   with module.terraform_state_backend.aws_s3_bucket.default,
│   on .terraform\modules\terraform_state_backend\main.tf line 111, in resource "aws_s3_bucket" "default":
│  111:   policy        = local.policy
│
│ Can't configure a value for "policy": its value will be decided automatically based on the result of applying this
│ configuration.

**Comment by apparentlymart** [comment]
Hi @minjas,

If you run the command `terraform providers` then it should tell you the individual provider requirements for each module in your configuration. Hopefully that will help narrow down which module has the `=> 4.0.0` constraint that was causing the conflict.

**Comment by minjas** [comment]
Hi @apparentlymart,

we have found out that **terraform-aws-modules/rds/aws** version was problem, so changing **version = "~> 3"** to  **version = "= 3.5.0"** solved issue in our case.

Apparently **version = "~> 3"** was translated to **>=4** in .terraform folder. Not sure why.

**Comment by apparentlymart** [comment]
Hi @minjas,

It sounds like you have found a resolution that works for you, so I'm going to close this.

Another possible resolution here would be to add a version constraint to your module block to select an earlier version of the module that does not require AWS provider v4.0 or later. I would suggest adding exact version constraints to any module block in your configuration that is calling a shared module from the public registry, because shared module authors can publish new major releases with breaking changes at any time, and so recording exactly the version you intend to use in your configuration will allow you to choose when to upgrade, rather than being immediately broken by those releases.

Thanks!
